### PR TITLE
fix(ci): add weekly k3k Calico sentinel

### LINKED ISSUE #6733: fix(ci): k3k-calico-host system test blocks 9 of 10 open PRs, and never runs on main
> 🤖 Generated by the Agentic Engineer

## Evidence

`🧪 System Test (Docker) (Vanilla, Docker, true, --name k3k-calico-host, true, K3s, Calico)` fails on
**9 of the 10 open pull requests** in this repository, including six Dependabot PRs that already have
auto-merge armed and therefore cannot complete. It is a single failure mode blocking the whole merge lane.

Representative failure — [job 98640219121](https://github.com/devantler-tech/ksail/actions/runs/33103257897/job/98640219121)
(PR #6725, 2026-08-27T20:04Z). The k3k container's own K3s bootstrap times out waiting for the node to
adopt the pod IP:

```
info "Waiting for Node IP to update to ${POD_IP}."
until kubectl get nodes -o wide 2>/dev/null | grep -q "${POD_IP}"; do
  ...
  if [ $count -gt 60 ]; then
    fatal "timed out waiting for node to change IP from $CURRENT_IP to $POD_IP"
```

→ `❌ K3s create FAILED (exit code: 1)`. The bound is 60 iterations × 2 s = **120 s**.

## Why "green on main" is not evidence here

⚠️ **`🧪 System Test (Docker)` is `skipped` on `main`** — confirmed on run
[33100039747](https://github.com/devantler-tech/ksail/actions/runs/33100039747), where it is the only
System Test entry and its conclusion is `skipped`. The job runs on pull requests and not on the default
branch, so a green `main` says nothing about this test's health and **must not be read as "the PRs
broke it"**. The defect may equally live on `main` with nothing there to catch it.

## Impact

- Nine open PRs cannot merge, including the Dependabot backlog (#6715, #6719, #6721, #6722, #6723, #6725).
- Because the test does not run on `main`, a regression here is invisible until it blocks every PR at once —
  which is what happened.

## What to determine first

1. Whether the timeout is a genuine regression or a **marginal** 120 s bound (the pattern in #6666's sibling
   diagnosis: a leg that always fails at the *same step* is not "flaky" even when its pass rate looks random).
2. Which of the 10 open PRs passes, and what distinguishes it.
3. Whether running this job on `main` (or a schedule) is the right fix for the detection gap, independent of
   the timeout itself.

## Acceptance criteria

- [ ] Root cause identified as either a real regression or an under-sized wait, with evidence.
- [ ] `k3k-calico-host` passes on a PR that previously failed it.
- [ ] The detection gap is addressed or explicitly declined with a reason: today nothing exercises this test
      on the default branch.

Size: medium. Likely needs a dedicated investigation.

### PR DESCRIPTION
**Pull request by devantler**
> 🤖 Generated by the Agentic Engineer

## Summary

- run the k3k + Calico compatibility path weekly on the default branch
- keep the scheduled check to one bounded system-test leg instead of the full matrix
- add a CI-harness contract test for cadence, scope, permissions, and cleanup

Fixes #6733

## Root cause

The `k3k-calico-host` leg only ran inside the pull-request matrix or by manual dispatch. Its dependencies can change behavior without a KSail diff, so the default branch had no recurring signal after merge. The runtime incompatibility originally reported by #6733 was fixed by #6709; this PR closes the remaining detection gap without scheduling the entire system-test matrix.

## Verification

RED at `5bb1a9256a11bd47a6869265ecde8c5b9b54cdc4`:

```text
go test ./internal/ciharness -run TestK3KCalicoSentinelRunsWeeklyOutsidePullRequests -count=1
open ../../.github/workflows/system-test-k3k-calico.yaml: no such file or directory
```

Initial GREEN at `03361da03fe695f089f406ed1df9cdd53d37facc`:

```text
go test -race ./internal/ciharness
ok github.com/devantler-tech/ksail/v7/internal/ciharness

go vet ./internal/ciharness
PASS

actionlint .github/workflows/system-test-k3k-calico.yaml
PASS

zizmor .github/workflows/system-test-k3k-calico.yaml
PASS
```

The workflow also self-validates when its own file changes. That contract was RED at `ed0418b1f3512874ffb3c7d1ab0447ca5a2e5431` because the pull-request path list was empty, then GREEN at the current head after adding the narrow workflow-file trigger. This makes the real sentinel run on this PR without adding it to unrelated pull requests.

The first substantive review exposed a credential edge case for fork/Dependabot pull requests. The workflow contract was RED at `03361da03fe695f089f406ed1df9cdd53d37facc` because the login checked only the username, PR self-runs received registry-token inputs, and the job requested package-write. It was GREEN at `86c71df92a1b37c045b6edd91efc6e2dabc75625`: PR runs are credential-free with `packages: read`, while non-PR runs retain optional registry authentication.

The restarted review then found that a manual dispatch from a non-default ref could still receive credentials and that the contract did not assert the manual trigger. Both cases were RED at `86c71df92a1b37c045b6edd91efc6e2dabc75625` and GREEN from `a8c81336cbb3148d579dfe62fc032dd107f6d00d`: credential inputs require the repository default branch, and the parsed-workflow contract asserts `workflow_dispatch`.

Repository lint then caught a long test function and an overlong condition. Those gates were RED at `a8c81336cbb3148d579dfe62fc032dd107f6d00d`, then GREEN at `6ff6354fe53ca7452b54ac508c72af151afb52ff` after a structural test-helper extraction and the repository formatter's exact wrap.

The next review identified a tag-ref collision: a manual dispatch from a tag whose short name matched the default branch could satisfy the name-only credential guard. All three credential-path assertions were RED at `6ff6354fe53ca7452b54ac508c72af151afb52ff`, then GREEN at the final `31d08d265565a3fcacd9f6b191f94f995a5ec810` after requiring `github.ref_type == 'branch'`. The focused contract also passed 10 consecutive runs.

The final current-head suite is 26 checks passed and zero failed, including build, race tests, coverage, vulnerability scanning, golangci-lint, MegaLinter, CodeQL, dependency review, zizmor, home isolation, provenance, and the required-check aggregate. CodeRabbit's restarted exact-head review reported no actionable issues after verifying the branch provenance guard and live result.

Local full-build and lint gates were unavailable: two full `go build` attempts exhausted the host filesystem while compiling the existing dependency graph; local `golangci-lint` panicked in its Go 1.26 analyzer on a file requiring Go 1.27; and `mega-linter-runner` is not installed. Current-head repository CI supplied those gates.

The real `k3k-calico-host` sentinel passed at the final exact head in [run 33166270876](https://github.com/devantler-tech/ksail/actions/runs/33166270876), including binary build, Helm/mirror cache restore, the credential-free PR k3k + Calico user path, diagnostics, logs, and always-run cleanup.

### CODE REVIEW

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer


@coderabbitai review

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

## Exact-head user-path evaluation

- **Head:** `03361da03fe695f089f406ed1df9cdd53d37facc`
- **Observed run:** [k3k + Calico Sentinel 33157394184](https://github.com/devantler-tech/ksail/actions/runs/33157394184)
- **What I exercised:** built KSail, initialized the vanilla Docker host, created the nested K3s cluster named `k3k-calico-host` with Calico, applied the podinfo overlay, tested Kubernetes-provider behavior, collected diagnostics, and ran cleanup.
- **User judgment:** the new scheduled surface successfully exercises the compatibility path it is meant to guard, and the bounded cleanup completed. This satisfies the real-user-path condition at this head.

The PR remains draft until its acknowledged external review returns a substantive current-head result.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

Resolved in `86c71df92a1b37c045b6edd91efc6e2dabc75625` after independently reproducing the credential edge case.

- Docker Hub login now requires a non-PR event, a configured username, and a non-empty step-level token.
- Pull-request self-validation receives no Docker Hub or GHCR credential and has only `packages: read`; scheduled/manual default-branch runs retain credentials.
- The workflow contract was RED for the missing token/event/permission guards, then GREEN after the change. `go test -race ./internal/ciharness`, `go vet ./internal/ciharness`, actionlint, and zizmor pass locally.

This also resolves the review summary's explicit package-write/registry-credential boundary concern. The pushed head will receive fresh CI, a real sentinel run, and a new exact-head review before promotion.

File: .github/workflows/system-test-k3k-calico.yaml [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        with:
+          docker-prune: "true"
+
+      - name: 🔐 Login to Docker Hub
+        if: ${{ vars.DOCKERHUB_USERNAME != '' }}
+        uses: docker/login-action@dbcb813823bdd20940b903addbd779551569679f # v4.6.0
+        with:
+          username: ${{ vars.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }} # zizmor: ignore[secrets-outside-env] - third-party action input, no environment scoping available
+
```
[the remark this answers is not in the document]
Comment (devantler) [reply .github/workflows/system-test-k3k-calico.yaml:52]: > 🤖 Generated by the Agentic Engineer

Resolved in `86c71df92a1b37c045b6edd91efc6e2dabc75625` after independently reproducing the credential edge case.

- Docker Hub login now requires a non-PR event, a configured username, and a non-empty step-level token.
- Pull-request self-validation receives no Docker Hub or GHCR credential and has only `packages: read`; scheduled/manual default-branch runs retain credentials.
- The workflow contract was RED for the missing token/event/permission guards, then GREEN after the change. `go test -race ./internal/ciharness`, `go vet ./internal/ciharness`, actionlint, and zizmor pass locally.

This also resolves the review summary's explicit package-write/registry-credential boundary concern. The pushed head will receive fresh CI, a real sentinel run, and a new exact-head review before promotion.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

## Exact-head user-path evaluation

- **Head:** `86c71df92a1b37c045b6edd91efc6e2dabc75625`
- **Observed run:** [k3k + Calico Sentinel 33159792173](https://github.com/devantler-tech/ksail/actions/runs/33159792173)
- **What I exercised:** the pull-request path built KSail and completed the vanilla Docker host → nested K3s `k3k-calico-host` + Calico → podinfo/Kubernetes-provider checks → diagnostics → cleanup sequence with PR registry credentials intentionally withheld.
- **User judgment:** the sentinel still exercises and completes the compatibility path under the hardened pull-request boundary; scheduled/manual default-branch runs retain optional registry authentication.

The PR remains draft until the remaining current-head checks and a restarted external review are green.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer


@coderabbitai review

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

Resolved both current-head findings in `a8c81336cbb3148d579dfe62fc032dd107f6d00d` after reproducing them as failing workflow-contract assertions.

- Docker Hub login and both registry-token inputs now require `github.ref_name == github.event.repository.default_branch`; manually dispatching a non-default ref receives no credentials.
- `workflow_dispatch` is modeled explicitly and asserted non-nil, so removing the manual trigger fails the contract test.
- The focused contract was RED for all missing guards, then GREEN after the change. `go test -race ./internal/ciharness`, `go vet ./internal/ciharness`, actionlint, and zizmor pass locally.

This reply resolves the inline security thread and records the specific fix for the outside-diff manual-trigger finding. Fresh exact-head CI, sentinel execution, and review remain required before promotion.

File: .github/workflows/system-test-k3k-calico.yaml [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+      - name: 🧹 Free disk space
+        uses: ./.github/actions/free-disk-space
+        with:
+          docker-prune: "true"
+
+      - name: 🔐 Login to Docker Hub
+        if: ${{ github.event_name != 'pull_request' && vars.DOCKERHUB_USERNAME != '' && env.DOCKERHUB_TOKEN != '' }}
+        env:
+          DOCKERHUB_TOKEN: ${{ secrets.DOCKERHUB_TOKEN }}
```
[the remark this answers is not in the document]
Comment (devantler) [reply .github/workflows/system-test-k3k-calico.yaml]: > 🤖 Generated by the Agentic Engineer

Resolved both current-head findings in `a8c81336cbb3148d579dfe62fc032dd107f6d00d` after reproducing them as failing workflow-contract assertions.

- Docker Hub login and both registry-token inputs now require `github.ref_name == github.event.repository.default_branch`; manually dispatching a non-default ref receives no credentials.
- `workflow_dispatch` is modeled explicitly and asserted non-nil, so removing the manual trigger fails the contract test.
- The focused contract was RED for all missing guards, then GREEN after the change. `go test -race ./internal/ciharness`, `go vet ./internal/ciharness`, actionlint, and zizmor pass locally.

This reply resolves the inline security thread and records the specific fix for the outside-diff manual-trigger finding. Fresh exact-head CI, sentinel execution, and review remain required before promotion.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

User evaluation at `6ff6354fe53ca7452b54ac508c72af151afb52ff`: I exercised the new workflow through its real pull-request entrypoint in [run 33163993924](https://github.com/devantler-tech/ksail/actions/runs/33163993924). As the sentinel's user, I observed the intended bounded path: the workflow built KSail, restored the Helm and mirror caches, created the nested K3s host, ran the exact `k3k-calico-host` Calico system test successfully, uploaded the normal logs, and completed the always-run cleanup. The pull-request run also skipped Docker Hub login, proving the credential-free self-validation state.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer


@coderabbitai review

File: .github/workflows/system-test-k3k-calico.yaml [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        uses: ./.github/actions/free-disk-space
+        with:
+          docker-prune: "true"
+
+      - name: 🔐 Login to Docker Hub
+        if: >-
+          ${{ github.event_name != 'pull_request'
+          && github.ref_name == github.event.repository.default_branch
+          && vars.DOCKERHUB_USERNAME != ''
+          && env.DOCKERHUB_TOKEN != '' }}
```
[the remark this answers is not in the document]
Comment (devantler) [reply .github/workflows/system-test-k3k-calico.yaml:45]: > 🤖 Generated by the Agentic Engineer

Resolved in `31d08d265565a3fcacd9f6b191f94f995a5ec810` after independently reproducing the tag-ref path.

- Docker Hub login and both registry-token expressions now require `github.ref_type == 'branch'` as well as the exact default-branch name.
- The parsed-workflow contract was RED at `6ff6354fe53ca7452b54ac508c72af151afb52ff` for all three missing guards, then GREEN after the change.
- `go test -race ./internal/ciharness`, `go vet ./internal/ciharness`, actionlint, zizmor, the 100-column check, and `git diff --check` pass locally.

The pushed head will receive fresh exact-head CI, sentinel execution, user evaluation, and review before promotion.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

User evaluation at `31d08d265565a3fcacd9f6b191f94f995a5ec810`: I exercised the security-fixed workflow through its real pull-request entrypoint in [run 33166270876](https://github.com/devantler-tech/ksail/actions/runs/33166270876). As the sentinel's user, I observed the intended bounded path: the workflow skipped Docker Hub login on the pull request, built KSail, restored the Helm and mirror caches, created the nested K3s host, ran the exact `k3k-calico-host` Calico system test successfully, uploaded the normal logs, and completed the always-run cleanup. This proves the credential-free self-validation state still works after requiring branch provenance for credentialed default-branch runs.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer


@coderabbitai review
